frontend/src/components/EditableCode.tsx

```diff
@@ -0,0 +1,67 @@
+import { useState } from 'react';
+import CodeEditor from '@uiw/react-textarea-code-editor';
+import ReactMarkdown from "react-markdown";
+import React from 'react';
+
+import Widget from './core/Widget';
+import { Icon } from "@blueprintjs/core";
+
+/**
+ * A component that allows toggling between editor and preview functionality.
+ * 
+ * Also accepts a save action that can be dispatched to publish the updates back to the server.
+ */
+const EditableCode = React.memo(({
+    initialCode,
+    code,
+    setCode,
+    langage,
+    className,
+}: {
+    initialCode: string,
+    code?: string,
+    setCode: (update: string) => void,
+    langage: string,
+    className?: string,
+}) => {
+    const [editing, setEditing] = useState(false);
+
+    const markdownWidgetButton = (
+        <div
+            onClick={() => setEditing(!editing)} 
+            className="p-1.5 rounded cursor-pointer hover:bg-slate-200/40">
+            {!editing ? 
```

Comment:
nit: logically simpler to just have `editing` be the root of the ternary instead of `!editing`